connect-location/src/main/java/com/ifttt/location/ConnectLocation.java

```diff
@@ -0,0 +1,97 @@
+
+public final class ConnectLocation implements ButtonStateChangeListener {
+
+    private static ConnectLocation INSTANCE = null;
+    private AwarenessGeofenceProvider awarenessGeofenceProvider;
+    private ConnectionApiClient connectionApiClient;
+
+    final static String FIELD_TYPE_LOCATION_ENTER = "LOCATION_ENTER";
+    final static String FIELD_TYPE_LOCATION_EXIT = "LOCATION_EXIT";
+    final static String FIELD_TYPE_LOCATION_ENTER_EXIT =
+            "LOCATION_ENTER_OR_EXIT";
+
+    private final static List<String> locationFieldTypesList = Arrays.asList(FIELD_TYPE_LOCATION_ENTER, FIELD_TYPE_LOCATION_EXIT, FIELD_TYPE_LOCATION_ENTER_EXIT);
+
+    public static synchronized ConnectLocation init(Context context, ConnectionApiClient apiClient) {
+        if (INSTANCE == null) {
+            INSTANCE = new ConnectLocation(context);
+        }
+        INSTANCE.connectionApiClient = apiClient;
+        return INSTANCE;
+    }
+
+    public static synchronized ConnectLocation init(Context context, CredentialsProvider credentialsProvider) {
+        if (INSTANCE == null) {
+            INSTANCE = new ConnectLocation(context);
+        }
+        ConnectionApiClient.Builder clientBuilder = new ConnectionApiClient.Builder(context);
+        INSTANCE.connectionApiClient = clientBuilder.build();
+        return INSTANCE;
+    }
+
+    private ConnectLocation(Context context) {
+        awarenessGeofenceProvider = new AwarenessGeofenceProvider(context);
+    }
+
+    public static void setUpWithConnectButton(ConnectButton connectButton) {
+        if (INSTANCE == null) {
+            throw new IllegalStateException("Connect Location is not initialized");
+        }
+        connectButton.addButtonStateChangeListener(INSTANCE);
+    }
+
+    @Override
+    public void onStateChanged(ConnectButtonState currentState, ConnectButtonState previousState, Connection connection) {
+        // Filter the location based enabled user feature fields.
+        // Using hard-coded data till API is ready
+
+        List<UserFeatureStep> userFeatureSteps = new ArrayList<>();
+
+        for (Feature feature :connection.features) {
+            if (feature.userFeatures != null) {
+                for (UserFeature userFeature : feature.userFeatures) {
+                    if (userFeature.userFeatureSteps != null) {
+                        userFeatureSteps.addAll(userFeature.userFeatureSteps);
+                    }
+                }
+            }
+        }
+
+        List<UserFeatureField> userFeatureFields = new ArrayList<>();
+
+        for (UserFeatureStep userFeatureStep: userFeatureSteps) {
+            if (userFeatureStep.fields != null) {
+                for (UserFeatureField userFeatureField : userFeatureStep.fields) {
+                    if (locationFieldTypesList.contains(userFeatureField.fieldType)) {
+                        userFeatureFields.add(userFeatureField);
+                    }
+                }
+            }
+        }
```

Comment:
The parsing logic may look a little repetitive because of the following:
1. The data that we are supposed to parse - i.e. lists of lists of lists...,
2. Api level constraints don't allow us to use `contains` or `List.stream().filter`
3. `addAll(...)` function input is annotated `RecentlyNotNull`, so there are additional nullability checks before we use that function.